File: Distributed/ebay.py
from candygram import *
import time
import collections

class ebayserver:
    def __init__(self):
        self.peers=set()
        self.price=0
        self.winner=self_()
        
    def launchthread(self,id):
        R = Receiver()
        R[str,Process,Any] = self.process, id, Message        
        R[str,Process,Any,Any] = self.process, id, Message        
        #R['register',Process,Any]=self.register, id, Message
        #R['bid',Process,Any,Any]=self.bid, id, Message
        for message in R:
          pass 

    #Candygram is broken
    def process(self,id,message):
        eval("self."+message[0])(id,message[1:])
    
    def register(self,id,message):
        FROM,ITEM=message
        print id, "recieved", message
        self.peers.add(FROM)
        FROM|("price","ITEM",self.winner,self.price)

    def bid(self,id,message):
        FROM,ITEM,PRICE=message
        print id, "recieved", message
        if PRICE>self.price:
            self.price=PRICE
            self.winner=FROM
            for p in self.peers:
                p |("price","ITEM",self.winner,self.price)

class Client:
    def __init__(self,server,maxprice):
        self.server=server
        self.maxprice=maxprice
    def launchthread(self,id):
        R = Receiver()
        R[str,Any,Process,Any]=self.process, id, Message
        self.server | ("register",self_(),"ITEM")
        for message in R:
          pass 

    def process(self,id,message):
        eval("self."+message[0])(id,message[1:])
    
    def price(self,id,message):
        item,winner,price = message
        print id, "recieved", message
        if winner!=self_() and price<self.maxprice:
           self.server | ("bid",self_(),"ITEM",price+1)




def launch(R,name):
    P = spawn(R.launchthread,name)
    R.proc = P
    print name, P
    return R
    
server = launch(ebayserver(),"S")
Client(1,2)
c1     = launch(Client(server.proc,20),"C1")
c2     = launch(Client(server.proc,30),"C2")

time.sleep(1)

File: Distributed/ebay.erl
-module(ebay).
-export([test/0,ebayserver/3,client/2]).
   
ebayserver(Price,Winner,Bidders) ->
 receive
	 {bid,ITEM,FROM,PRICE} -> 
		 if 
		   PRICE>Price ->
             lists:map(fun(B) -> B!{price,ITEM,FROM,PRICE} end,Bidders),
		     ebayserver(PRICE,FROM,Bidders);
	       PRICE=<Price ->
			   ebayserver(Price,Winner,Bidders)
	     end;
	 {registerforprice,ITEM,FROM} -> 
		 FROM!{price,ITEM,Winner,Price},
		 ebayserver(Price,Winner,[FROM]++Bidders)
		 
 end.

client(Server,Maxprice) ->
 receive
    {price,ITEM,Winner,Price} -> 
		io:format("C<~w> winner ~w price ~w~n",[self(),Winner,Price]),
		S=self(),
	    if 
		  Winner/=S ->
			  if Price<Maxprice ->
					  Server!{bid,ITEM,S,Price+1},
				      client(Server,Maxprice);
				 Price>=Maxprice ->
				      client(Server,Maxprice)
		      end;
		  Winner==S ->
			  client(Server,Maxprice)
	    end
 end.

test() ->
  S = spawn(ebay,ebayserver,[0,self(),[]]),
  C1 = spawn(ebay,client,[S,21]),
  C2 = spawn(ebay,client,[S,30]),
  S!{registerforprice,0,C1},
  S!{registerforprice,0,C2},
  io:format("Done.~n",[]).
  

